main.c

```diff
@@ -1867,31 +2013,58 @@ void editorProcessKeypress(int c) {
 #ifndef EMSYS_CUA
 	case CTRL('z'):
 #endif //EMSYS_CUA
+	{
+
 		for (int i = 0; i < rept; i++) {
-			bufr->cy = bufr->rowoff;
-			int times = (E.screenrows / E.nwindows) - 4;
-			while (times--)
-				editorMoveCursor(bufr, ARROW_UP);
+			int overlap = 2;
+			// Move the cursor as needed
+			bufr->cx = 0;
+			if (bufr->cy >= win->rowoff + overlap) {
+				bufr->cy = win->rowoff + 1;
+			}
+
+			// Move the window offset
+			win->rowoff -= (win->height - overlap);
+			if (win->rowoff < 0)
+				win->rowoff = 0;
 		}
-		break;
+
+		editorScroll();
+	} break;
 	case PAGE_DOWN:
 #ifndef EMSYS_CUA
 	case CTRL('v'):
 #endif //EMSYS_CUA
 		for (int i = 0; i < rept; i++) {
-			bufr->cy = bufr->rowoff + E.screenrows - 1;
-			if (bufr->cy > bufr->numrows)
-				bufr->cy = bufr->numrows;
-			int times = (E.screenrows / E.nwindows) - 4;
-			while (times--)
-				editorMoveCursor(bufr, ARROW_DOWN);
+			int overlap = 2;
+			// move the cursor as needed
+			bufr->cx = 0;
+
+			if (bufr->cy < win->rowoff + win->height - overlap) {
+				bufr->cy = win->rowoff + win->height - overlap;
+			}
+			// Move the window offset
+			win->rowoff += (win->height - overlap);
+			if (win->rowoff < 0)
+				win->rowoff = 0;
+
+			editorScroll();
 		}
 		break;
 	case BEG_OF_FILE:
 		bufr->cy = 0;
 		bufr->cx = 0;
-		bufr->rowoff = 0;
 		break;
+	case CUSTOM_INFO_MESSAGE: {
+		int winIdx = windowFocusedIdx(&E);
+		struct editorWindow *win = E.windows[winIdx];
+		struct editorBuffer *buf = win->buf;
+
+		editorSetStatusMessage(
+			"(buf->cx%d,cy%d) (win->scx%d,scy%d) win->height=%d screenrows=%d, screencols=%d",
+			buf->cx, buf->cy, win->scx, win->scy, win->height,
+			E.screenrows, E.screencols);
+	} break;
```

Comment:
Once more for luck :grin: 

```suggestion
	case CUSTOM_INFO_MESSAGE:
		int winIdx = windowFocusedIdx(&E);
		struct editorWindow *win = E.windows[winIdx];
		struct editorBuffer *buf = win->buf;

		editorSetStatusMessage(
			"(buf->cx%d,cy%d) (win->scx%d,scy%d) win->height=%d screenrows=%d, screencols=%d",
			buf->cx, buf->cy, win->scx, win->scy, win->height,
			E.screenrows, E.screencols);
		break;
```